lib/features/popular/data/repositories/movie_repository.dart

```diff
@@ -0,0 +1,94 @@
+import 'package:either_dart/either.dart';
+import 'package:hooks_riverpod/hooks_riverpod.dart';
+import 'package:movie_app/common/data/api_client.dart';
+import 'package:movie_app/common/data/providers.dart';
+import 'package:movie_app/common/utils/constants.dart';
+import 'package:movie_app/features/popular/data/mappers/movie_entity_mapper.dart';
+import 'package:movie_app/features/popular/data/models/movie_response.dart';
+import 'package:movie_app/features/popular/data/repositories/genre_repository.dart';
+import 'package:movie_app/features/popular/domain/entities/movie.dart';
+import 'package:movie_app/features/popular/domain/providers/all_genres_provider.dart';
+import 'package:movie_app/features/popular/domain/providers/total_pages_provider.dart';
+import 'package:movie_app/generated/l10n.dart';
+import 'package:q_architecture/q_architecture.dart';
+
+final movieRepositoryProvider = Provider<MovieRepository>(
+  (ref) => MovieRepositoryImpl(
+    ref,
+    ref.watch(apiClientProvider),
+    ref.watch(genreRepositoryProvider),
+    ref.watch(
+      movieEntityMapperProvider,
+    ),
+  ),
+  name: 'Movie repository provider',
+);
+
+abstract interface class MovieRepository {
+  EitherFailureOr<List<Movie>> getMovies(int page);
+}
+
+class MovieRepositoryImpl implements MovieRepository {
+  final Ref ref;
+  final ApiClient _apiClient;
+  final GenreRepository _genreRepository;
+  final EntityMapper<Movie, MovieResponse> _movieMapper;
+
+  MovieRepositoryImpl(
+    this.ref,
+    this._apiClient,
+    this._genreRepository,
+    this._movieMapper,
+  );
+
+  @override
+  EitherFailureOr<List<Movie>> getMovies(int page) async {
+    final movies = <Movie>[];
+    final genreMap = ref.read(allGenresProvider.notifier);
+
+    try {
+      final response = await _apiClient.getMovies(
+        kBearerToken,
+        kApiLanguage,
+        page,
+      );
+
+      ref.read(totalPagesProvider.notifier).state = response.totalPages;
```

Comment:
Ovdje umjesto da imaš totalPagesProvider bolje vraćati iz getMovies umjesto liste filmova, cijeli MovieResponseWrapper pretvoren u entity pa onda unutra uz listu imaš i informaciju o trenutnom pageu, totalPages, totalResults i samu listu filmova. Nema potrebe imati dodati provider za ovo, veća je mogućnost pogreške na taj način.